README.md
# feathers-offline-log

[![Build Status](https://travis-ci.org/feathersjs/feathers-offline-log.png?branch=master)](https://travis-ci.org/feathersjs/feathers-offline-log)
[![Code Climate](https://codeclimate.com/github/feathersjs/feathers-offline-log/badges/gpa.svg)](https://codeclimate.com/github/feathersjs/feathers-offline-log)
[![Test Coverage](https://codeclimate.com/github/feathersjs/feathers-offline-log/badges/coverage.svg)](https://codeclimate.com/github/feathersjs/feathers-offline-log/coverage)
[![Dependency Status](https://img.shields.io/david/feathersjs/feathers-offline-log.svg?style=flat-square)](https://david-dm.org/feathersjs/feathers-offline-log)
[![Download Status](https://img.shields.io/npm/dm/feathers-offline-log.svg?style=flat-square)](https://www.npmjs.com/package/feathers-offline-log)

| Performant, persistent client-side transaction log.

Browser client logs are kept in (preferred order) IndexedDB, WebSQL or localStorage.
Log entries are added in an efficient manner, whether the log is small or very large.
JavaScript objects are also handled in an efficient manner.

> **ProTip:** Used for feathers-offline's optimistic mutation log,
which is updated while the client is disconnected.

The cache stringifies an added object just once,
eliminating the multiple stringifications of alternative designs.
The cache bundles logs into chunks, whose maximum size you can specify (0.5 meg default).
This avoids both the repetitive persisting of large blobs
and the inefficiencies of writing individual log items.

## Installation

```
npm install feathers-offline-log --save
```

## Documentation

- **cache = new Cache(storageHandler, options)** - Configure a new cache.
    - `storageHandler` (required, module) - Module responsible for persistence.
        - `.../browser-storage` - Browser support. See example below.
    - `options` (optional) - Options for the cache.
        - `chunkMaxLen` (optional, number, default 500 000) - Maximum chunk size in chars.
        - `sep` (optional, char, default ',') - Separator char between log entries.
        Objects are always separated with ','.
- **cache.config(options)** - Configure the storage handler for the cache.
The options depend on the storage handler.
    - `.../browser-storage` - See options for
    [localforage](http://localforage.github.io/localForage/)
        - `name` (optional, string, default 'offline') - Namespace for storage.
        Typically the app name.
        - `collection` (optional, string, default 'log') - Namespace within `name`.
        Typically the name of the collection or table name.
- **cache.add(str)** - Add `str` to the logs.
- **cache.addObj(obj)** - Stringifyy `obj` and add it to the logs.
- **cache.getOldestChunk()** - Returns the oldest chunk of logs.
If it consists solely of added objects,
then `JSON.stringify('[' + chunk + ']')` will return an array of POJO.
- **cache.removeOldestChunk()** - Remove the oldest chunk from the log.
- **cache.clear()** - Remove all chunks.

## Example

```js
const browserStorage = require('feathers-offline-log/browser-storage');
const Cache = require('feathers-offline-log');

const cache = new Cache(browserStorage, { chunkMaxLen: 500000 }); // logs stored in 0.5 meg chunks
cache.config({ name: 'myApp', collection: 'people' })
  .then(() => cache.addObj(obj1))
  .then(() => cache.addObj(obj2))
  .then(() => cache.addObj(obj3))

  .then(() => cache.getOldestChunk()) 
  .then(chunk => console.log(JSON.parse(`[${chunk}]`))) 
  .then(() => cache.removeOldestChunk()) 
```

The included browser storage abstraction for
[localforage](http://localforage.github.io/localForage/)
may be used standalone.

```js
const browserStorage = require('feathers-offline-log/browser-storage');

browserStorage.config({ name: 'myApp', collection: 'people' })
  .then(() => browserStorage.setItem('john', ' ... ')) // full key is myApp_people/john
  .then(() => browserStorage.getItem('john'))
  .then(str => browserStorage.removeItem('john'))
  .then(() => browserStorage.clear())
  .then(() => browserStorage.length())
  .then(numb => browserStorage.keys())
  .then(array => browserStorage.iterate((value, key, iterationNumber) => { ... }));
````

## License

Copyright (c) 2017

Licensed under the [MIT license](LICENSE).


CHANGELOG.md
# Change Log

## [v0.0.1](https://github.com/feathersjs/feathers-offline-log/tree/v0.0.1) (2017-07-09)
**Closed issues:**

- Action required: Greenkeeper could not be activated 🚨 [\#1](https://github.com/feathersjs/feathers-offline-log/issues/1)

**Merged pull requests:**

- Renamed instanceName to collection [\#6](https://github.com/feathersjs/feathers-offline-log/pull/6) ([eddyystop](https://github.com/eddyystop))
- Improved codeclimate score [\#5](https://github.com/feathersjs/feathers-offline-log/pull/5) ([eddyystop](https://github.com/eddyystop))
- Changed storageBrowser to browserStorage [\#4](https://github.com/feathersjs/feathers-offline-log/pull/4) ([eddyystop](https://github.com/eddyystop))
- Greenkeeper/initial [\#3](https://github.com/feathersjs/feathers-offline-log/pull/3) ([eddyystop](https://github.com/eddyystop))
- Fixed tests [\#2](https://github.com/feathersjs/feathers-offline-log/pull/2) ([eddyystop](https://github.com/eddyystop))



\* *This Change Log was automatically generated by [github_changelog_generator](https://github.com/skywinder/Github-Changelog-Generator)*

test/cache.test.js

const assert = require('chai').assert;

const browserStorage = require('../src/browser-storage');
const Cache = require('../src');

let cache;

describe('cache', () => {
  let objs;
  let lens;
  
  beforeEach(() => {
    objs = [
      makeObj(),
      makeObj(),
      makeObj(),
    ];
    
    lens = objs.map((obj, i) => JSON.stringify(objs[i]).length);
  
    cache = new Cache(browserStorage, { chunkMaxLen: lens[1] + lens[2] + 50 });
    return cache.config()
      .then(() => cache.clear())
      .then(() => cache.config());
  });
  
  it('adds 1 obj, gets oldest chunk', () => {
    return cache.addObj(objs[0])
      
      .then(() => cache._getChunkKeys())
      .then(keys => assert.deepEqual(keys, ['_0']))
      
      .then(() => cache.getOldestChunk())
      .then(chunk => {
        assert.deepEqual(JSON.parse(`[${chunk}]`), [objs[0]]);
      });
  });
  
  it('adds 2 obj', () => {
    return cache.addObj(objs[0])
      .then(() => cache.addObj(objs[1]))
  
      .then(() => cache._getChunkKeys())
      .then(keys => assert.deepEqual(keys, ['_0']))
      
      .then(() => cache.getOldestChunk())
      .then(chunk => {
        assert.deepEqual(JSON.parse(`[${chunk}]`), [objs[0], objs[1]]);
      });
  });
  
  it('starts a new chunk', () => {
    return cache.addObj(objs[0])
      .then(() => cache.addObj(objs[1]))
      .then(() => cache.addObj(objs[2]))
      
      .then(() => cache._getChunkKeys())
      .then(keys => assert.deepEqual(keys, ['_0', '_1']))
  });
  
  it('removes a chunk', () => {
    return cache.addObj(objs[0])
      .then(() => cache.addObj(objs[1]))
      .then(() => cache.addObj(objs[2]))
      
      .then(() => cache.removeOldestChunk())
      
      .then(() => cache._getChunkKeys())
      .then(keys => assert.deepEqual(keys, ['_1']))
      
      .then(() => cache.getOldestChunk())
      .then(chunk => assert.deepEqual(JSON.parse(`[${chunk}]`), [objs[2]]));
  });
});

function makeObj() {
  const unique = Math.random();
  
  return {
    name: { first: 'eddyy', middle: 'boomer', last: 'stop', unique },
    address: { line1: '5 Vallet of Shadows',
      line2: 'upper floor',
      city: 'Orgrimmar',
      country: 'Kalimdor',
      planet: 'Azeroth',
      unique },
    messages: [
      'Death to the Alliance!',
      'Vanilla WoW was best! Burning Crusade was darn good.',
      unique
    ]
  };
}


test/helpers/init.js

global.localStorage = require('humble-localstorage'); // eslint-disable-line


test/stringify.test.js


function make(count, seed) {
  const obj = [count, seed];
  
  for (let i = 0; i < count; i++) {
    const unique = seed + count + i;
    
    obj.push({
      name: { first: 'eddyy', middle: 'boomer', last: 'stop', unique },
      address: { line1: '5 Vallet of Shadows', line2: 'upper floor',
        city: 'Orgrimmar', country: 'Kalimdor', planet: 'Azeroth', unique },
      messages: [
        'Death to the Alliance!',
        'Vanilla WoW was best! Burning Crusade was pretty good.',
        unique
      ]
    });
  }
  
  return obj;
}

const loops = 10;
const times = [];
let str;

for (let loop = 0; loop < loops; loop++) {
  const obj = make(20000, Math.random());
  let start = process.hrtime();
  
  str = JSON.stringify(obj);
  
  let end = process.hrtime(start);
  times.push(end[1]/1000000);
}

const avg = times.reduce((total, curr) => total + curr, 0) / loops;
console.log('%d megs, average %d ms', str.length/1000000, avg);
console.log(times);

src/cache.js

import makeDebug from 'debug';
const debug = makeDebug('log-cache');

export default class Cache {
  constructor(storageHandler, options = {}) {
    this._chunkMaxLen = options.chunkMaxLen || 500000; // 0.5 meg
    this._sep = options.sep || ',';
    this._storageHandler = storageHandler;
    this._currChunkKey = '';
    this._currChunk = '';
  }
  
  config(options = {}) {
    debug('config started', options);
    
    return this._storageHandler.config(options)
      .then(() => this._getChunkKeys())
      .then(keys => {
        if (keys.length) {
          this._currChunkKey = this._incrKey(keys[keys.length - 1]);
        } else {
          this._currChunkKey = '_0';
        }
        
        debug('config ended. _currChunkKey=', this._currChunkKey);
      });
  }
  
  add(str, sep) {
    debug('add started', this._currChunkKey);
    sep = sep || this._sep;
  
    return new Promise(resolve => {
      if (this._currChunk.length + str.length <= this._chunkMaxLen) {
        return resolve();
      }
  
      const promise = this._flushCurrChunk()
        .then(() => {
          this._currChunkKey = this._incrKey(this._currChunkKey);
          this._currChunk = '';
        });
  
      return resolve(promise);
    })
      .then(() => {
        this._currChunk = `${this._currChunk}${this._currChunk.length ? sep : ''}${str}`;
        this._flushCurrChunk();
        debug('add ended. length', this._currChunk.length);
      })
  }
  
  addObj(obj) {
    debug('addObj entered');
    return this.add(JSON.stringify(obj), ',');
  }

  getOldestChunk() {
    debug('getOldestChunk entered');
    return this._getChunkKeys()
      .then(keys => this._storageHandler.getItem(keys[0]));
  }
  
  removeOldestChunk() {
    debug('removeOldestChunk entered');
    return this._getChunkKeys()
      .then(keys => this._storageHandler.returnItem(keys[0]));
  }
  
  clear() {
    debug('clear entered');
    return this._storageHandler.clear();
  }
  
  _getChunkKeys() {
    debug('_getChunkKeys entered');
    return this._storageHandler.keys()
      .then(keys => keys.filter(key => key.charAt(0) === '_').sort());
  }
  
  _incrKey(lastKey) {
    return `_${(parseInt(lastKey.substr(1), 10) + 1) + ''}`;
  }
  
  _flushCurrChunk() {
    debug('_flushCurrChunk entered. key', this._currChunkKey);
    return this._storageHandler.setItem(this._currChunkKey, this._currChunk);
  }
};


src/browser-storage.js

const localforage = require('localforage');

module.exports = {
  _logger: null,

  config (options = {}) {
    options.collection = options.collection || 'log';
    options.name = options.name || 'offline';

    localforage.config(options);
    const _logger = this._logger = localforage.createInstance({ name: options.collection });

    this.setItem = _logger.setItem;
    this.getItem = _logger.getItem;
    this.removeItem = _logger.removeItem;
    this.clear = _logger.clear;
    this.length = _logger.length;
    this.keys = _logger.keys;
    this.iterate = _logger.iterate;

    return Promise.resolve();
  },

  setObj (key, obj) {
    return this.setItem(key, JSON.stringify(obj));
  },

  getObj (key) {
    return this.getItem(key)
      .then(str => JSON.parse(str));
  }
};


src/storage-browser.js

const localforage = require('localforage');

module.exports = {
  _logger: null,

  config (options = {}) {
    options.instanceName = options.instanceName || 'log';
    options.name = options.name || 'offline';

    localforage.config(options);
    const _logger = this._logger = localforage.createInstance({ name: options.instanceName });

    this.setItem = _logger.setItem;
    this.getItem = _logger.getItem;
    this.removeItem = _logger.removeItem;
    this.clear = _logger.clear;
    this.length = _logger.length;
    this.keys = _logger.keys;
    this.iterate = _logger.iterate;

    return Promise.resolve();
  },

  setObj (key, obj) {
    return this.setItem(key, JSON.stringify(obj));
  },

  getObj (key) {
    return this.getItem(key)
      .then(str => JSON.parse(str));
  }
};


src/index.js

import makeDebug from 'debug';
const debug = makeDebug('log-cache');

export default class Cache {
  constructor (storageHandler, options = {}) {
    this._chunkMaxLen = options.chunkMaxLen || 500000; // 0.5 meg
    this._sep = options.sep || ',';
    this._storageHandler = storageHandler;
    this._currChunkKey = '';
    this._currChunk = '';
  }

  config (options = {}) {
    debug('config started', options);

    return this._storageHandler.config(options)
      .then(() => this._getChunkKeys())
      .then(keys => {
        if (keys.length) {
          this._currChunkKey = this._incrKey(keys[keys.length - 1]);
        } else {
          this._currChunkKey = '_0';
        }

        debug('config ended. _currChunkKey=', this._currChunkKey);
      });
  }

  add (str, sep) {
    debug('add started', this._currChunkKey);
    sep = sep || this._sep;

    return new Promise(resolve => {
      if (this._currChunk.length + str.length <= this._chunkMaxLen) {
        return resolve();
      }

      const promise = this._flushCurrChunk()
        .then(() => {
          this._currChunkKey = this._incrKey(this._currChunkKey);
          this._currChunk = '';
        });

      return resolve(promise);
    })
      .then(() => {
        this._currChunk = `${this._currChunk}${this._currChunk.length ? sep : ''}${str}`;
        this._flushCurrChunk();
        debug('add ended. length', this._currChunk.length);
      });
  }

  addObj (obj) {
    debug('addObj entered');
    return this.add(JSON.stringify(obj), ',');
  }

  getOldestChunk () {
    debug('getOldestChunk entered');
    return this._getChunkKeys()
      .then(keys => keys.length ? this._storageHandler.getItem(keys[0]) : undefined);
  }

  removeOldestChunk () {
    debug('removeOldestChunk entered');
    return this._getChunkKeys()
      .then(keys => keys.length ? this._storageHandler.removeItem(keys[0]) : undefined);
  }

  clear () {
    debug('clear entered');
    return this._storageHandler.clear();
  }

  _getChunkKeys () {
    debug('_getChunkKeys entered');
    return this._storageHandler.keys()
      .then(keys => keys.filter(key => key.charAt(0) === '_').sort());
  }

  _incrKey (lastKey) {
    return `_${(parseInt(lastKey.substr(1), 10) + 1) + ''}`;
  }

  _flushCurrChunk () {
    debug('_flushCurrChunk entered. key', this._currChunkKey);
    return this._storageHandler.setItem(this._currChunkKey, this._currChunk);
  }
}
